src/main/java/com/xiaomi/infra/pegasus/rpc/async/TableHandler.java

```diff
@@ -307,16 +346,44 @@ void call(final ClientRequestRound round, final int tryId) {
     final TableConfiguration tableConfig = tableConfig_.get();
     final ReplicaConfiguration handle =
         tableConfig.replicas.get(round.getOperator().get_gpid().get_pidx());
-    if (handle.session != null) {
-      handle.session.asyncSend(
+
+    // send request to primary session
+    if (handle.primarySession != null) {
+      // if it's not write operation and backup request is enabled, schedule to send to secondaries
+      if (!round.operator.isWrite && isBackupRequestEnabled()) {
+        round.backupRequstTask =
+            executor_.schedule(
+                new Runnable() {
+                  @Override
+                  public void run() {
+                    for (ReplicaSession session : handle.secondarySessions) {
```

Comment:
Your implementation is to emit two backup RPCs to both secondaries. This is not consistent with what you designed.
Here we should choose randomly one secondary only:

>  but then falls back on sending a secondary request after the first request has been outstanding for more than the 95th-percentile(or 99th-percent line, etc) expected latency.

Reply:
I suggest take turns choosing one secondary but randomly.

Reply:
I suggest take turns but not randomly.

Reply:
I've been waiting for this discussion for a long time. Because if we send to all of these secondaries, we can get the fastest reply from all of these replicas, but the server load will increase(but not too much, because we send to secondaries only when the primary has been outstanding for more than the 95th percentline). So I think this is a trade off.

Reply:
I remember paper is `one secondary`, or we can add switch to choose `one` or `two`

Reply:
I will do performance tests to compare these two ways.

Reply:
There is a compromise between the two implementations. We can send a backup request to one of these secondaries(random selection) after delay `backup_request_delay_time`. And wait for the response of them to arrive. If they donot come back after `backup_request_delay_time`, send to another secondary. In the other words, we emit request to the last secondary after delay `2 * backup_request_delay_time`.

Reply:
@Shuo-Jia 

Doubling the secondaries for backup-request also doubling the risk of cascading failure, when Pegasus has been overwhelmed with a high load (10K QPS for example), the additional load brought by backup-request will be ~20K QPS. Because all the requests fail to complete in expected P99 latency under this load.

Furthermore, we should consider how to prevent cascading failure brought by backup request. My idea is to statistic the ratio of "backup request" and "normal request", and

Reply:
> statistic the ratio of "backup request" and "normal request", and if
> the ratio is higher than 5%

`statistic` may be a slightly complicated, for example, you need add qps counter such as server-side and calc the ratio every time?  if you want the  ratio of `backup request` is below 5%, can we send `backup request `by proportional sampling?

Reply:
Sure. It's merely an idea.